### ISSUE TITLE
subarray and prange problem

### ISSUE THREAD

**Issue by ninegua** [IntelLabs/numba#13]
The following code seems to fail with a weird error:

```
@njit(parallel=True)
def f(n):
  b = np.zeros((n, n))
  a = b[0]
  for j in prange(n):
      a[j] = j + 1
  return b.sum()
```

It fails at:

```
  File "numba/npyufunc/parfor.py", line 49, in _lower_parfor_parallel
    assert parfor.params
```

Any idea? @ehsantn 

Ideally, I would like the following to work, which is how nested array comprehension eventually becomes:

```
@njit(parallel=True)
def f(n):
  b = np.zeros((n, n))
  for i in prange(n):
    a = b[i]
    for j in prange(n):
      a[j] = i + j
  return b.sum()
```

With `range` it works correctly, but the result is wrong with `prange`.

**Comment by ehsantn** [comment]
Seems like the problem is that remove dead doesn't know `a` and `b` are aliased. We haven't had any subarray case before, so we should think about handling them properly in general.

**Comment by ninegua** [comment]
I think `remove_dead` can be a bit more conservative, e.g., arrays that is being written should not be considered dead.

**Comment by ehsantn** [comment]
Removing dead arrays has proven very useful actually. Handling aliases shouldn't be that difficult. I'll push a fix to the array-comprehension branch ASAP.

**Comment by ninegua** [comment]
Another failed case:

```
@njit(parallel=True)
def g(n):
  b = np.zeros((n, n, n))
  for i in prange(n):
    a = b[i]
    for j in prange(n):
      c = a[j]
      for k in prange(n):
        c[k] = i + j + k
  return b.sum()
```

This creates a situation where the gufunc produced from outer prange goes through another round of type checking, and yet, itself contains nested parfors. Variables defined inside parfor body got assigned `Del` nodes outside parfor body, which fails type checking.

**Comment by ehsantn** [comment]
This revealed a deep and scary bug which is fixed with 5e2caacba54334823983b57a89646a20841f158b now.

**Comment by ninegua** [comment]
thanks for the fix! The above code works now. However, nested parallel array comprehension with 3 level now compiles without error, but still returns `nan`:

```
@njit(parallel=True)
def f(n):
   a = np.array([ [ [z + y * x for z in range(n)]  for y in range(n) ] for x in range(n) ])
  return np.sum(a)
```

With `parallel=False` it works. The untyped IR looks similar to the above 3-level for-loop, which works but this doesn't. So there must be something else going on.

**Comment by ninegua** [comment]
I pushed the test cases for parallel comprehension, now this 3 level test case fails at `test_comp_nest_with_array_3`.

**Comment by ehsantn** [comment]
The `prange` translation for this code seems correct. I don't see obvious issues in the optimized IR either. We need to look at it more closely.

**Comment by ehsantn** [comment]
works with sequential lowering so gufunc translation/optimization might have a problem.

**Comment by DrTodd13** [comment]
Fixed in 084e60f..d7b1051.
